1.5/Source/PawnStorages/PawnStorages/Farm/Comps/CompFarmStorage.cs

```diff
@@ -24,9 +24,35 @@ public class CompFarmStorage : CompPawnStorage
         public float NutritionRequiredPerDay() => compAssignable.AssignedPawns.Sum(animal =>
             SimplifiedPastureNutritionSimulator.NutritionConsumedPerDay(animal.def, animal.ageTracker.CurLifeStage));
 
-        public new void StorePawn(Pawn pawn)
+        public IEnumerable<PawnKindDef> HeldPawnTypes => this.innerContainer.innerList.Select(p => p.kindDef).Distinct();
+
+        public Dialog_AutoSlaughter.AnimalCountRecord CountForType(ThingDef def)
+        {
+            List<Pawn> pawns = innerContainer.innerList.Where(p => p.kindDef.race == def).ToList();
+            int total = pawns.Count;
+            int male = pawns.Count(p => p.gender == Gender.Male && p.ageTracker.Adult);
+            int maleYoung = pawns.Count(p => p.gender == Gender.Male && !p.ageTracker.Adult);
+            int female = pawns.Count(p => p.gender == Gender.Female && p.ageTracker.Adult);
+            int femaleYoung = pawns.Count(p => p.gender == Gender.Female && !p.ageTracker.Adult);
+
+            return new Dialog_AutoSlaughter.AnimalCountRecord(total, male, maleYoung, female, femaleYoung, 0, 0);
+        }
+
+        public void StorePawn(Pawn pawn, bool despawnFirst = true)
         {
-            pawn.DeSpawn();
+            if(despawnFirst)
+                pawn.DeSpawn();
+
+            if (innerContainer.Count >= MaxStoredPawns())
+            {
+                Messages.Message("PS_StorageFull".Translate(parent.LabelCap, pawn.LabelCap), (Thing) pawn, MessageTypeDefOf.NeutralEvent);
+
+                PawnComponentsUtility.AddComponentsForSpawn(pawn);
+                compAssignable?.TryUnassignPawn(pawn);
+                GenDrop.TryDropSpawn(pawn, parent.Position, parent.Map, ThingPlaceMode.Near, out Thing _);
```

Comment:
I think this has a potential bug; if despawnFirst is false and the storage is full it'll dupe the pawn.
Why would you ever not despawn first?

Reply:
You wouldn't despawn first in the case where we've generated a new pawn from breeding, but not spawned it yet. 
In this case, when a pawn is "born" it's generated, but not spawned, then passed to the `Notify_PawnBorn` method.
This passes it to `StorePawn` which either stores it, or spawns it. That's the only codepath where `despawnFirst` will be false.